#11 송금 서비스 DB 모델링 및 ERD 작성

donggi-lee-bit opened the pull request:
안녕하세요 @f-lab-maverick 😄
송금 서비스에 대한 DB 모델링 진행하였고, 모델링 결과는 ERD로 작성하였습니다.

주요 테이블, 컬럼 정보는 다음과 같습니다 (구현 방식이 결정나지 않은 부분은 제외하고 변하지 않을 것들을 우선하여 모델링하였습니다)
`remittance_request`: 송금 요청 테이블
- `sender_id`: 송금인 ID
- `receiver_id`: 수취인 ID
- `status`: 송금 요청 상태
  - `PENDING`: 요청 대기 상태
  - `ACCEPTED`: 요청 수락 상태
  - `EXPIRED`: 만료된 상태
  - `CANCELLED`: 요청 취소 상태
- `amount`: 송금액
- `created_at`: 송금 요청 시간
- `expired_at`: 송금 요청 만료 시간

`remittance_status_history`: 송금 상태 로그 테이블
- `request_id`: 송금 요청 테이블 FK
- `status`: 송금 요청 상태
  - `PENDING`: 요청 대기 상태
  - `ACCEPTED`: 요청 수락 상태
  - `EXPIRED`: 만료된 상태
  - `CANCELLED`: 요청 취소 상태
- `created_at`: 로그 생성 시간

`remittance_history`: 송금, 수취 내역 테이블
- `remittance_request_id`: 송금 요청 테이블 FK
- `member_id`: 송금, 수취 대상자 ID
- `description`: 송금시 작성한 메모
- `amount`: 송금액 
- `balance`: 잔액
- `created_at`: 송금 내역 생성 시간

### 고민 부분

- 송금과 수취 내역을 따로 관리하기에는 중복되는 부분이 많아 통합한 테이블로 관리하게끔 했습니다. 지금과 같이 유사한 데이터를 통합하여 관리하게 되면 추후 분리해야할 상황이 생긴다면 분리가 어려울 것 같은데 이는 트레이드 오프가 되는건지.. 정답이 있는 문제인건지 궁금합니다.
- 유효한 ID를 보장하기 위해 연관 테이블의 컬럼에 모두 FK를 설정했습니다. 많은 FK 설정은 데이터 쓰기 성능에도 안좋은 영향을 줄 것 같은데 FK를 모든 연관 컬럼에 설정하는 것이 데이터 무결성을 보장하는 최선의 방법인지 궁금합니다.

donggi-lee-bit committed:
테이블 생성 쿼리 파일 추가

donggi-lee-bit committed:
리드미에 ERD 추가

f-lab-maverick commented:
![image](https://github.com/user-attachments/assets/0555126c-558b-414a-96e7-b595dfd34a3d)
위 ERD를 기반으로 질문&피드백 남기겠습니다.

1. auth 테이블을 별도로 분리하신 이유가 있나요? 패스워드를 별도 테이블에 저장하는 이유가 있을지 궁금합니다. member테이블에 같이 보관하지 않는다면 유저를 조회할 때 마다 패스워드를 찾기 위해 auth 테이블과의 join이 필요하지 않을까요?
2. `remittance_status_history`테이블의 경우 송금요청에 대한 히스토리를 보관하는 히스토리성 테이블로 보입니다. 이런 히스토리성 테이블은 join하지 않더라도 히스토리로만 내용을 어느정도 확인할 수 있게 만드는것이 좋습니다. 히스토리는 그 자체로 변경되지 않아야하는 불변성이 요구되는데요, join을 통해 원본테이블을 조회하도록 하면 원본테이블의 내용이 변할때마다 히스토리의 내용도 변하기 때문입니다. 또한 히스토리는 송금요청이 유효하지 않더라도 저장되어야 합니다. 송금기록 없이도 히스토리가 쌓일 수 있는 구조가 필요해보입니다.
3. member_remittance_history 는 왜 필요하다고 생각하셨나요? member - remittance_hitsory를 다대다관계로 정의해서 중간테이블이 나왔는데요, 이미 remittance_hitsory 테이블은 member_id를 가지고 있습니다. 다대다로 정의하신 이유가 있으실까요? 별도로 필요하지 않다면 member_remittance_history 테이블을 제거하고 remittance_history의 member_id를 활용해 직접 연결하는 방식을 고려할 수 있습니다.
4. remittance_history의 balance는 어떤 컬럼인가요?

donggi-lee-bit commented:
> 1. auth 테이블을 별도로 분리하신 이유가 있나요? 패스워드를 별도 테이블에 저장하는 이유가 있을지 궁금합니다. member테이블에 같이 보관하지 않는다면 유저를 조회할 때 마다 패스워드를 찾기 위해 auth 테이블과의 join이 필요하지 않을까요?

맞습니다. auth 테이블은 인증과 관련된 데이터를 포함하고자 했습니다. 이는 추후 로그인 구현이 확정되었을 때 관련하여 컬럼이 추가될 예정입니다! 말씀하신 것처럼 password는 member 테이블에서 같이 관리되는게 맞습니다. 해당 부분 수정하도록 하겠습니다!

donggi-lee-bit commented:
> 2. `remittance_status_history`테이블의 경우 송금요청에 대한 히스토리를 보관하는 히스토리성 테이블로 보입니다. 이런 히스토리성 테이블은 join하지 않더라도 히스토리로만 내용을 어느정도 확인할 수 있게 만드는것이 좋습니다. 히스토리는 그 자체로 변경되지 않아야하는 불변성이 요구되는데요, join을 통해 원본테이블을 조회하도록 하면 원본테이블의 내용이 변할때마다 히스토리의 내용도 변하기 때문입니다. 

설계 당시 join을 통해 원본 테이블을 조회하면 원본 테이블의 데이터가 변할 수 있을 거라는 점을 인지하지 못했습니다. 😂 `remittance_status_history` 테이블만으로 히스토리를 보여주기 위해서는 추가로 sender_id, receiver_id, amount 컬럼을 추가해주면 될까요? 그리고 해당 테이블에 설정된 외래키는 제거하는 게 좋을까요?

> 또한 히스토리는 송금요청이 유효하지 않더라도 저장되어야 합니다. 송금기록 없이도 히스토리가 쌓일 수 있는 구조가 필요해보입니다.

송금 요청이 유효하지 않더라도 히스토리가 저장되어야 하는 이유가 궁금합니다! 비정상 요청을 확인하는 용도로 사용되는걸까요?

donggi-lee-bit commented:
> 3. member_remittance_history 는 왜 필요하다고 생각하셨나요? member - remittance_hitsory를 다대다관계로 정의해서 중간테이블이 나왔는데요, 이미 remittance_hitsory 테이블은 member_id를 가지고 있습니다. 다대다로 정의하신 이유가 있으실까요? 별도로 필요하지 않다면 member_remittance_history 테이블을 제거하고 remittance_history의 member_id를 활용해 직접 연결하는 방식을 고려할 수 있습니다.

맞습니다.. `remittance_history` 테이블의 member_id를 사용하면 되기 때문에 `member_remittance_history` 테이블은 제거하겠습니다!

donggi-lee-bit commented:
> 4. remittance_history의 balance는 어떤 컬럼인가요?

balance 컬럼은 잔액을 의미하는 컬럼입니다. 해당 컬럼은 잔액 테이블을 따로 관리한다면 제거될 것 같습니다. 하지만 아직 어떤 방식으로 잔액 데이터를 관리할지 고민입니다 🤔

f-lab-maverick commented:
Q. 설계 당시 join을 통해 원본 테이블을 조회하면 원본 테이블의 데이터가 변할 수 있을 거라는 점을 인지하지 못했습니다. 😂 remittance_status_history 테이블만으로 히스토리를 보여주기 위해서는 추가로 sender_id, receiver_id, amount 컬럼을 추가해주면 될까요? 그리고 해당 테이블에 설정된 외래키는 제거하는 게 좋을까요?
A. 히스토리는 그 자체로 데이터를 알 수 있게 많은양의 컬럼을 추가하셔도 좋습니다. 외래키또한 원본 데이터를 추적할 수 있는 용도로 사용할 수 있으므로, 같이 저장해주셔도 됩니다. 히스토리테이블은 일반적으로 정규화를 하지 않습니다 :)

Q.송금 요청이 유효하지 않더라도 히스토리가 저장되어야 하는 이유가 궁금합니다! 비정상 요청을 확인하는 용도로 사용되는걸까요?
A. 어떤 비즈니스 로직을 구현하느냐에 따라 다르지만, 일반적으로 송금요청 그 자체만으로도 히스토리를 남기는 경우가 있습니다.  히스토리를 저장할 때 그 목적은 데이터변경 또는 요청을 추적하기 위함인데요, 사용자가 정상적으로 요청하였지만 동시성이슈등으로 블락되어 타임아웃이 나거나 한 경우 유효하지 않은 요청이지만 이 요청 자체는 히스토리에 남아있어야 이후 오류 제보등을 받았을 때 추적하기 편합니다.

수정한부분을 추가적으로 push해주신 후 다시 리뷰 진행하시죠~

donggi-lee-bit committed:
테이블 설계 리뷰 반영

- password 컬럼을 member 테이블로 이동
- 히스토리 테이블을 원본 테이블과 join하지 않고도 히스토리 내용을 파악할 수 있도록 컬럼 추가
- member_remittance_history 테이블 제거

donggi-lee-bit commented:
수정 사항 반영하였습니다! 🙂 @f-lab-maverick

f-lab-maverick commented:
https://github.com/user-attachments/assets/2bde8a13-c1de-4f5b-a1d9-7b36731e012b

f-lab-maverick reviewed (changes requested):
자세한 리뷰사항은 멘토링 스크립트를 참고해주세요.
수정후 다시 리뷰요청해주세요~

donggi-lee-bit committed:
JOIN 없이 히스토리 정보를 조회할 수 있도록 DB 모델링 개선

- 송금 히스토리, 송금 상태 히스토리 정보를 원본 테이블과 JOIN 하지 않고 테이블의 컬럼으로만 히스토리 정보를 알 수 있도록 컬럼 추가
- 그 외
  - status, email, password 컬럼 사이즈 수정

donggi-lee-bit commented:
스크립트를 참고해 리뷰해주신 내용 반영했습니다! @f-lab-maverick 

- `auth` 테이블 제거
- 히스토리 테이블과 원본 테이블 사이의 외래키 제약 조건 제거
- 컬럼 사이즈 조정
  - `email` (100 -> **300**)
  - `password` ( 100 -> **300**)
  - `status` (20 -> **50**)
-  `remittance_status_history` 테이블에 만료 시점, 수취 시점을 기록하기 위해 `expired_at`, `accepted_at` 컬럼 추가

f-lab-maverick reviewed (approved):
고생하셨습니다. 머지해주셔도 좋습니다.
